add base protocol upgrade spec (#874)
vegaprotocol/specs: protocol/0075-PLUP_protocol_upgrades.md

The solution employed at the time was to:
- implement a hotfix
- keep the hotfix behind a guard until a given time, once the time is reached network parameters would be dispatched.
- distribute the code to the validator so they can test it.
- then decide of an actual date which would give a week for the validator to asynchronously update their node.

This upgrade went quite smoothly without any incident, although it was possible as the fix was located in a single place, and wasn't impacting much of the state
(meaning no changes in the protobuf files were required for example.)

## Protocol upgrades mechanism

One of the main challenge with upgrading the network both asynchronously and without a restart / downtime is
the management of the state, and possible incompatible changes between the state of two version of the protocol.

The following describes the general workflow of upgrades: 
1. Vega continuesly make new releases available 
2. Validators can suggest to the network an upgrade with a given release tag and a block height for the upgrade to take place

> 3. If a proposal gets enough votes (`protocol.upgrade.proposal.requiredMajority` of the validators, not stake) at the given height vega will take a snapshot automatically and stop processing further blocks until restarted by the vega-visor, the vega process manager. 

Reviewer: I am guessing this was discussed - but does it make sense to count validators not something like "own + delegated stake" or "tendermint weight"? Because the purpose is for the chain to continue post upgrade - so I'd think we need not only majority but the "right ones". Thoughts?

Reviewer 2: @barnabee - please can you review this and confirm if we should leave as is, or factor in the tendermint weight?

Reviewer 2: @jgsbennett - once we have a decision here can you please add an AC accordingly?

Reviewer 2: Decision in spec review meeting: We will remain as is and use count and not factor in score/weights for this milestone

Reviewer 3: Net param is actually validators.vote.required . Rename here and in any ACs that mention.

}

```

## Acceptance criteria

### Invalid proposals - Rejections
   - A network with 5 validators 
   - (<a name="0075-PLUP-001" href="#0075-PLUP-001">0075-PLUP-001</a>) Validator proposes a protocol upgrade to an invalid tag (https://semver.org/) - should result in an error
   - (<a name="0075-PLUP-002" href="#0075-PLUP-002">0075-PLUP-002</a>) Validator proposes a protocol upgrade on a block height preceding the current block - should result in an error
   - (<a name="0075-PLUP-003" href="#0075-PLUP-003">0075-PLUP-003</a>) Propose and enact a version downgrade
   - (<a name="0075-PLUP-004" href="#0075-PLUP-004">0075-PLUP-004</a>) Non-validator attempts to propose upgrade
   - (<a name="0075-PLUP-005" href="#0075-PLUP-005">0075-PLUP-005</a>) Ersatz validator attempts to propose upgrade
 
### Block height validation
   Proposal will not be accepted as valid if validator
   -  (<a name="0075-PLUP-006" href="#0075-PLUP-006">0075-PLUP-006</a>) Proposes a negative upgrade block
   -  (<a name="0075-PLUP-007" href="#0075-PLUP-007">0075-PLUP-007</a>) Proposes a 0 upgrade block
   -  (<a name="0075-PLUP-008" href="#0075-PLUP-008">0075-PLUP-008</a>) Proposes (string/other upgrade block)

>    -  (<a name="0075-PLUP-009" href="#0075-PLUP-009">0075-PLUP-009</a>) Proposes without supplying a block height

Reviewer: Should we add an AC that the block number has to be strictly greater than the current block?

Reviewer 2: @jgsbennett - please can you add this as an AC?

### VISOR 

>    - (<a name="0075-PLUP-010" href="#0075-PLUP-010">0075-PLUP-010</a>) Can be seen to automatically download the latest version for install when available at the source location when file meets the format criteria defined

Reviewer: The latest  version and the one in an accepted proposal may be different. Should we add an AC that the vegavisor downloads all the versions? Or all the ones in active proposal? Or is that not needed / intended behavior?

>    - (<a name="0075-PLUP-010" href="#0075-PLUP-010">0075-PLUP-010</a>) Can be seen to automatically download the latest version for install when available at the source location when file meets the format criteria defined

Reviewer 2: ```suggestion
   - (<a name="0075-PLUP-010" href="#0075-PLUP-010">0075-PLUP-010</a>) Can be seen to automatically download the tagged version proposed for install when available at the source location when file meets the format criteria defined
```

   - (<a name="0075-PLUP-011" href="#0075-PLUP-011">0075-PLUP-011</a>) Visor automatically upgrades validators to proposed version if requiredMajority has been reached
   
### Epochs 

>    - (<a name="0075-PLUP-012" href="#0075-PLUP-012">0075-PLUP-012</a>) Proposing an upgrade block which ought to be the end of an epoch

Reviewer: ... should have what result? Is it just we want to test that if an upgrade block and epoch end coincide then all goes well?

Reviewer 3: We should come back up in the new epoch, and rewards should have been paid out.

Reviewer 2: Try to test this with null blockchain. If this proves to be very complex we can consider testing this later: ```suggestion
   - (<a name="0075-COSMICELEVATOR-012" href="#0075-COSMICELEVATOR-012">0075-COSMICELEVATOR-012</a>) Proposing an upgrade block which ought to be the end of an epoch
```

   - (<a name="0075-PLUP-017" href="#0075-PLUP-017">0075-PLUP-017</a>) Events are emitted for all proposals which fail to reach required majority when target block is reached
   - (<a name="0075-PLUP-018" href="#0075-PLUP-018">0075-PLUP-018</a>) When majority reached during the process of upgrading, those validators which didnt propose will stop producing blocks
   - (<a name="0075-PLUP-019" href="#0075-PLUP-019">0075-PLUP-019</a>) Proposals for multiple versions at same block height will be rejected if majority has not been reached, network continues with the current running version 
   - (<a name="0075-PLUP-020" href="#0075-PLUP-020">0075-PLUP-020</a>) Propose with a validator which is moved to Ersatz by the time the upgrade is enacted. If there are 5 validators, 3 vote yes, 2 vote no: One of the yes voters is kicked in favour of a new one, leaving the vote at 2-2 so the upgrade should not happen as counting votes happens at block height only

### Multiple proposals (<a name="0075-PLUP-021" href="#0075-PLUP-021">0075-PLUP-021</a>)
   - If multiple proposals are submitted from a validator before the block heights are reached then only the last proposal is considered

> ## Spam (<a name="0075-PLUP-022" href="#0075-PLUP-022">0075-PLUP-022</a>) 

Reviewer: ```suggestion
## Spam (<a name="0075-COSMICELEVATOR-022" href="#0075-COSMICELEVATOR-022">0075-COSMICELEVATOR-022</a>) 
```

   - Excessive numbers of proposals from a single validator within an epoch should be detected and rejected - (Future requirement)
   
## Snapshots

>    - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post validator joining they should be immediately allowed to propose and included in the overall total count

Reviewer: "should be immediately allowed to propose" what does propose refer to here? Propose blocks? Propose new upgrades?

Reviewer 3: I think the latter.

>    - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post validator joining they should be immediately allowed to propose and included in the overall total count

Reviewer 2: ```suggestion
   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post a validator becoming a tendermint validator they should be immediately allowed to propose an upgrade and be included in the overall total count
```

>    - (<a name="0075-PLUP-024" href="#0075-PLUP-024">0075-PLUP-024</a>) Validators attempting to join during upgrade will be rejected if protocol versions differ

Reviewer: Can we define "during" more precisely? Does during mean "proposal to update at block 1000 has been submitted, we are now at block 500 so no-one can join"? 

Or does "during" mean really we got to block 1000 the chain stopped while nodes restart and when it restarts it will again accept transactions and we create block 1001?

Reviewer 3: Again, I think the latter.

Reviewer 3: Drop this AC. We think it's just another specific case of the "validator joins with the wrong version" scenario.

>    - (<a name="0075-PLUP-025" href="#0075-PLUP-025">0075-PLUP-025</a>) Ensure that required majority is not met when enough validators join between validator proposals and target block

Reviewer: What does this mean?

Reviewer 3: I don't know for sure what was originally intended, but I might imagine:
1: 5 validators, 3 vote for upgrade to version V on block B.
2: Prior to that, 2 more validators join, giving total 7 validators. Block B passes without upgrade (if new validators do not vote).
(Intention is that required majority is met in step 1, but not in step 2. I haven't played with that field yet, may also require votes from the new validators to achieve this case.)

Reviewer 3: Clarify AC.

   - (<a name="0075-PLUP-026" href="#0075-PLUP-026">0075-PLUP-026</a>) Node starting from snapshot which has a proposal at a given block, ensure during replay when the block height is reached a new version is loaded and also post load an upgrade takes place at target block.

> ## Checkpoints

Reviewer: ```suggestion
## LNL Checkpoints
```

   - (<a name="0075-PLUP-027" href="#0075-PLUP-027">0075-PLUP-027</a>) Validator proposals should not be stored in the checkpoints and restored into the network
   - (<a name="0075-PLUP-028" href="#0075-PLUP-028">0075-PLUP-028</a>) Upgrade will not occur after a post checkpoint restore until new proposals are made and block height reached
   
## API
   - (<a name="0075-PLUP-029" href="#0075-PLUP-029">0075-PLUP-029</a>) An datanode API should be available to provide information on the upcoming confirmed proposal including total proposals/block details/versions
   

### Successful upgrade  (<a name="0075-PLUP-030" href="#0075-PLUP-030">0075-PLUP-030</a>) 
  - A new release is made available, and is successfully deployed
  - Setup a network with 5 validators running version x
  - Have 4 validator submit request to upgrade to release >x at block height 1000
  - At the end of block height 1000 a snapshot is taken and vega is stopped by the vegavisor
  - All nodes are starting from the snapshot of block 1000 and the network resumes with version >x
 
 
 ### Failing consensus (<a name="0075-PLUP-031" href="#0075-PLUP-031">0075-PLUP-031</a>)

>   - Upgrade takes place at block N. Minimum number of validators -1 restart with correct version. One validator restarts with previous version. Consensus is not achieved. Start another validator with the correct version, consensus is achieved

Reviewer: Regardless of what own stake + delegations they have?

Reviewer 3: Step 1:
Restart the network with a number of validators whose voting power is equal to less than two thirds. Consensus is not achieved.
Step 2:
Restart one of the validators (whose voting power would take the total voting power over the two thirds threshold) with the wrong version. Consensus should still not be achieved.
Step 3:
Restart validator from step 2 with the correct version. Consensus is achieved.

  - 5 validator network. Upgrade takes places at block N. Start 3 validators immediately. Allow several blocks to pass. Start two remaining validators. (All validators continue to work) - no blocks producing as 3 validators do not have enough weight - need 70% weight to produce blocks

>   - Weighting threshold (API TM) is not reached after restart, network does not produce blocks

Reviewer 3: Drop

>   - Upgrade takes place, but insufficient validators are restored for 1, 5, 10, minutes. Validators which are restored immediately patiently wait for consensus to be achieved, and then blocks continue  - concensus acheived

Reviewer 3: Add an AC for the mainnet test case. https://github.com/vegaprotocol/devops-infra/issues/1522

Reviewer 3: Add an AC for "submit a proposal with version == current version and a valid block height retracts any existing proposals"
Add an AC for "submit an invalid proposal does not overwrite any existing proposals" i.e: Set up an upgrade set to happen at block height H. One validator (whose changing vote would cause the upgrade to not happen) submits an invalid upgrade proposal (i.e: invalid version). Proposal rejected. Upgrade still happens.

   string vega_release_tag = 2;
   // the release tag for the data-node binary
   string data_node_release_tag = 3;
   // tendermint validators that have agreed to the upgrade
  repeated string approvers = 4;
  // the status of the proposal
  ProtocolUpgradeProposalStatus status = 5;
}

```

## Acceptance criteria

### Invalid proposals - Rejections (<a name="0075-PLUP-001" href="#0075-PLUP-001">0075-PLUP-001</a>)
   - A network with 5 validators 
   - Validator 1 proposes a protocol upgrade to an invalid tag (https://semver.org/) - should result in an error
   - Validator 2 proposes a protocol upgrade on a block height preceding the current block - should result in an error
   - Propose and enact a version downgrade 
   - Non-validator attempts to propose upgrade
   - Ersatz validator attempts to propose upgrade
 
### Block height validation (<a name="0075-PLUP-002" href="#0075-PLUP-002">0075-PLUP-002</a>)
   Proposal will not be accepted as valid if validator
   -  Proposes an upgrade block in the past - should be rejected / failed check - sign via wallet / check wallet
   -  Proposes a negative upgrade block
   -  Proposes a 0 upgrade block

>    -  Proposes (string/other upgrade block)

Reviewer 3: @jeremyletang I think the python/protos code I've used so far tries to enforce an integer type. To what extent is it possible to send an incorrect type over the wire? i.e: By hacking together funky transactions, like with the work Muthu is doing.

Author: yea well just try to build transaction yourself would be what you can do really.

>    -  Proposes with supply a block height

Reviewer 3: Is this supposed to be "without supplying" ?

Reviewer 4: yes

### VISOR (<a name="0075-PLUP-005" href="#0075-PLUP-005">0075-PLUP-005</a>) 
   - Can be seen to automatically download the latest version for install when available at the source location when file meets the format criteria defined
   - Visor automatically upgrades validators to proposed version if requiredMajority has been reached
   
### Epochs (<a name="0075-PLUP-006" href="#0075-PLUP-006">0075-PLUP-006</a>) 
   - Proposing an upgrade block which ought to be the end of an epoch

>    - Propose an upgrade block which should result in a new network in the same epoch

Reviewer 3: These are maybe 3 different test cases, which one do we mark with the AC? Similar deal for several of the above... Our current system seems to assume one python test case per AC... For _some_ of these I could theoretically cram them into one case, but for some of the others it doesn't necessarily make sense.

   -  (<a name="0075-PLUP-006" href="#0075-PLUP-006">0075-PLUP-006</a>) Proposes a negative upgrade block
   -  (<a name="0075-PLUP-007" href="#0075-PLUP-007">0075-PLUP-007</a>) Proposes a 0 upgrade block
   -  (<a name="0075-PLUP-008" href="#0075-PLUP-008">0075-PLUP-008</a>) Proposes (string/other upgrade block)
   -  (<a name="0075-PLUP-009" href="#0075-PLUP-009">0075-PLUP-009</a>) Proposes without supplying a block height

### VISOR 
   - (<a name="0075-PLUP-010" href="#0075-PLUP-010">0075-PLUP-010</a>) Can be seen to automatically download the tagged version proposed for install when available at the source location when file meets the format criteria defined
   - (<a name="0075-PLUP-011" href="#0075-PLUP-011">0075-PLUP-011</a>) Visor automatically upgrades validators to proposed version if required majority has been reached
   
### Epochs 
   - (<a name="(0075-COSMICELEVATOR-012)" href="#(0075-COSMICELEVATOR-012)">(0075-COSMICELEVATOR-012)</a>) Proposing an upgrade block which ought to be the end of an epoch. After upgrade takes place, confirm rewards are distributed, any pending delegations take effect, and validator joining/leaving takes effect.

>    - (<a name="0075-PLUP-013" href="#0075-PLUP-013">0075-PLUP-013</a>) Propose an upgrade block which should result in a new network in the same epoch

Reviewer: ```suggestion
   - (<a name="0075-PLUP-013" href="#0075-PLUP-013">0075-PLUP-013</a>) Propose an upgrade block which should result in a network running new code version in the same epoch.
```

   - (<a name="0075-PLUP-014" href="#0075-PLUP-014">0075-PLUP-014</a>) Ensure end of epoch processes still run after restore e.g reward calculation and distributions

> ### Required Majority

Reviewer: ```suggestion
### Required Majority
For the purposes of protocol upgrade each validator that participates in consensus has one vote. Required majority is set by `validators.vote.required network parameter`. 
```

   - (<a name="0075-PLUP-015" href="#0075-PLUP-015">0075-PLUP-015</a>) Counting proposal votes to check if required majority has been reached occurs when any proposed target block has been reached

>    - (<a name="0075-PLUP-016" href="#0075-PLUP-016">0075-PLUP-016</a>) Only active network validators proposals are counted when any proposed target block has been reached 

Reviewer: ```suggestion
   - (<a name="0075-PLUP-016" href="#0075-PLUP-016">0075-PLUP-016</a>) Only proposals from validators participating in consensus are counted when any proposed target block has been reached. 
```

   - (<a name="0075-PLUP-017" href="#0075-PLUP-017">0075-PLUP-017</a>) Events are emitted for all proposals which fail to reach required majority when target block is reached
   - (<a name="0075-PLUP-018" href="#0075-PLUP-018">0075-PLUP-018</a>) When majority reached during the process of upgrading, those validators which didnt propose will stop producing blocks
   - (<a name="0075-PLUP-019" href="#0075-PLUP-019">0075-PLUP-019</a>) Proposals for multiple versions at same block height will be rejected if majority has not been reached, network continues with the current running version 
   - (<a name="0075-PLUP-020" href="#0075-PLUP-020">0075-PLUP-020</a>) Propose with a validator which is moved to Ersatz by the time the upgrade is enacted. If there are 5 validators, 3 vote yes, 2 vote no: One of the yes voters is kicked in favour of a new one, leaving the vote at 2-2 so the upgrade should not happen as counting votes happens at block height only

>    - (<a name="0075-PLUP-036" href="#0075-PLUP-036">0075-PLUP-036</a>) Changing validators.vote.required network parameter to a value above two thirds is respected.

Reviewer: ```suggestion
   - (<a name="0075-PLUP-036" href="#0075-PLUP-036">0075-PLUP-036</a>) Changing `validators.vote.required` network parameter to a value above two thirds is respected.
```

>    - (<a name="0075-PLUP-037" href="#0075-PLUP-037">0075-PLUP-037</a>) The value of validators.vote.required is checked at upgrade block, i.e: vote on a proposal with all validators, then change the validators.vote.required net param before upgrade block, to a higher value, which would cause the upgrade to be rejected. Upgrade fails.

Reviewer: ```suggestion
   - (<a name="0075-PLUP-037" href="#0075-PLUP-037">0075-PLUP-037</a>) The value of `validators.vote.required` is checked at upgrade block, i.e: vote on a proposal with all validators, then change the `validators.vote.required` net param before upgrade block, to a higher value, which would cause the upgrade to be rejected. Upgrade fails.
```

### Multiple proposals (<a name="0075-PLUP-021" href="#0075-PLUP-021">0075-PLUP-021</a>)
   - If multiple proposals are submitted from a validator before the block heights are reached then only the last proposal is considered

## Spam (<a name="0075-COSMICELEVATOR-022" href="#0075-COSMICELEVATOR-022">0075-COSMICELEVATOR-022</a>) 
   - Excessive numbers of proposals from a single validator within an epoch should be detected and rejected - (Future requirement)
   
## Snapshots

>    - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post a validator becoming a tendermint validator they should be immediately allowed to propose an upgrade and be included in the overall total count

Reviewer: ```suggestion
   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post a validator becoming a consensus-participating validator they should be immediately allowed to propose an upgrade and be included in the overall total count
```

   
## API
   - (<a name="0075-PLUP-028" href="#0075-PLUP-028">0075-PLUP-028</a>) An datanode API should be available to provide information on the upcoming confirmed proposal including total proposals/block details/versions
   

### Successful upgrade  (<a name="0075-PLUP-029" href="#0075-PLUP-029">0075-PLUP-029</a>)
  - A new release is made available, and is successfully deployed
  - Setup a network with 5 validators running version x
  - Have 4 validator submit request to upgrade to release >x at block height 1000
  - At the end of block height 1000 a snapshot is taken and vega is stopped by the vegavisor
  - All nodes are starting from the snapshot of block 1000 and the network resumes with version >x
 
 
### Failing consensus
  - (<a name="0075-PLUP-030" href="#0075-PLUP-030">0075-PLUP-030</a>) Upgrade takes place at block N. Restart with a number of validators whose voting power is <= two thirds. Restart one more validator whose voting power would take the total voting power >= two thirds, with an incorrect version. Consensus is not achieved. Now restart that validator with the correct version. Consensus is achieved.

>   - (<a name="0075-PLUP-031" href="#0075-PLUP-031">0075-PLUP-031</a>) 5 validator network. Upgrade takes places at block N. Start 3 validators immediately. Allow several blocks to pass. Start two remaining validators. (All validators continue to work) - no blocks producing as 3 validators do not have enough weight - need 70% weight to produce blocks

Reviewer: Allow several blocks to pass. ... no blocks producing as 3 validators do not have enough weight  

This is confusing. Do we mean allow *time* to pass first.

Reviewer 3: Yes, I never saw that. I'll correct it.
